frame/elections-phragmen/src/lib.rs

```diff
@@ -406,55 +524,93 @@ decl_module! {
-			// around. The main intention is that reading all of the candidates, members and runners
-			// from storage is expensive. Furthermore, we know (soft proof) that they are always
-			// mutually exclusive. Hence, we try one, and only then decode more storage.
-
-			if let Ok(_replacement) = Self::remove_and_replace_member(&who) {
-				T::Currency::unreserve(&who, T::CandidacyBond::get());
-				Self::deposit_event(RawEvent::MemberRenounced(who.clone()));
-
-				// safety guard to make sure we do only one arm. Better to read runners later.
-				return Ok(());
-			}
-
-			let mut runners_up_with_stake = Self::runners_up();
-			if let Some(index) = runners_up_with_stake.iter()
-				.position(|(ref r, ref _s)| r == &who)
-			{
-				runners_up_with_stake.remove(index);
-				// unreserve the bond
-				T::Currency::unreserve(&who, T::CandidacyBond::get());
-				// update storage.
-				<RunnersUp<T>>::put(runners_up_with_stake);
-				// safety guard to make sure we do only one arm. Better to read runners later.
-				return Ok(());
-			}
-
-			let mut candidates = Self::candidates();
-			if let Ok(index) = candidates.binary_search(&who) {
-				candidates.remove(index);
-				// unreserve the bond
-				T::Currency::unreserve(&who, T::CandidacyBond::get());
-				// update storage.
-				<Candidates<T>>::put(candidates);
-				// safety guard to make sure we do only one arm. Better to read runners later.
-				return Ok(());
+		/// <weight>
+		/// If a candidate is renouncing:
+		/// 	Base weight: 17.28 µs
+		/// 	Complexity of candidate_count: 0.235 µs
+		/// 	State reads:
+		/// 		- Candidates
+		/// 		- [AccountBalance(who) (unreserve)]
+		/// 	State writes:
+		/// 		- Candidates
+		/// 		- [AccountBalance(who) (unreserve)]
+		/// If member is renouncing:
+		/// 	Base weight: 46.25 µs
+		/// 	State reads:
+		/// 		- Members, RunnersUp (remove_and_replace_member),
+		/// 		- [AccountData(who) (unreserve)]
+		/// 	State writes:
+		/// 		- Members, RunnersUp (remove_and_replace_member),
+		/// 		- [AccountData(who) (unreserve)]
+		/// If runner is renouncing:
+		/// 	Base weight: 46.25 µs
+		/// 	State reads:
+		/// 		- RunnersUp (remove_and_replace_member),
+		/// 		- [AccountData(who) (unreserve)]
+		/// 	State writes:
+		/// 		- RunnersUp (remove_and_replace_member),
+		/// 		- [AccountData(who) (unreserve)]
+		///
+		/// Weight note: The call into changeMembers need to be accounted for.
+		/// </weight>
+		#[weight =  match *renouncing {
+			Renouncing::Candidate(count) => {
+				18 * WEIGHT_PER_MICROS +
+				(count as Weight) * 235 * WEIGHT_PER_NANOS +
+				T::DbWeight::get().reads_writes(1, 1)
+			},
+			Renouncing::Member => {
+				46 * WEIGHT_PER_MICROS +
+				T::DbWeight::get().reads_writes(2, 2)
+			},
+			Renouncing::RunnerUp => {
+				46 * WEIGHT_PER_MICROS +
+				T::DbWeight::get().reads_writes(1, 1)
 			}
-
-			Err(Error::<T>::InvalidOrigin)?
+		}]
+		fn renounce_candidacy(origin, renouncing: Renouncing) {
```

Comment:
This would be 3 functions if I were designing the API

Reply:
It would not require this renouncing type for example, and would reduce the size of the tx by removing one parameter

Reply:
Store for another day?

Reply:
indeed